File: web/frontend/src/api/notifications.ts
import request from '../utils/request';

export interface NotificationItem {
  id: number;
  createdAt: number;
  scheduleId: string | null;
  taskId: string | null;
  taskName: string | null;
  engine: string | null;
  status: string;
  title: string;
  summary: string | null;
  readAt: number | null;
}

export interface UnreadCountResponse {
  count: number;
}

export async function fetchNotifications(options?: {
  limit?: number;
  unreadOnly?: boolean;
}): Promise<NotificationItem[]> {
  const params = new URLSearchParams();
  if (options?.limit !== undefined) {
    params.set('limit', String(options.limit));
  }
  if (options?.unreadOnly !== undefined) {
    params.set('unread_only', String(options.unreadOnly));
  }
  const query = params.toString();
  return request<NotificationItem[]>(`/api/notifications${query ? `?${query}` : ''}`);
}

export async function fetchUnreadCount(): Promise<number> {
  const res = await request<UnreadCountResponse>('/api/notifications/unread-count');
  return res.count;
}

export async function markNotificationRead(id: number): Promise<void> {
  await request<{ status: string }>(`/api/notifications/${id}/read`, {
    method: 'POST',
  });
}

export async function markAllNotificationsRead(): Promise<number> {
  const res = await request<{ marked: number }>('/api/notifications/read-all', {
    method: 'POST',
  });
  return res.marked;
}

File: core/web/routers/notifications.py
from __future__ import annotations

from fastapi import APIRouter, Path, Query, Request

from core.services.run_store import RunStore
from core.web.case_convert import ProtocolModel, wire
from core.web.routers import tasks as tasks_router

router = APIRouter(prefix="/api", tags=["notifications"])


def _get_run_store(request: Request) -> RunStore:
    if hasattr(request.app.state, "run_store") and request.app.state.run_store is not None:
        return request.app.state.run_store
    return tasks_router._runner._run_store


class NotificationItem(ProtocolModel):
    id: int
    created_at: float
    schedule_id: str | None = None
    task_id: str | None = None
    task_name: str | None = None
    engine: str | None = None
    status: str
    title: str
    summary: str | None = None
    read_at: float | None = None


@router.get("/notifications")
def list_notifications(
    request: Request,
    limit: int = Query(50, le=100),
    unread_only: bool = Query(False),
) -> list[dict]:
    """Lists recent notifications (most recent first)."""
    store = _get_run_store(request)
    rows = store.list_notifications(limit=limit, unread_only=unread_only)
    result = []
    for row in rows:
        item = wire(NotificationItem(**row))
        item["task_id"] = row.get("task_id")
        item["schedule_id"] = row.get("schedule_id")
        item["task_name"] = row.get("task_name")
        item["created_at"] = row.get("created_at")
        item["read_at"] = row.get("read_at")
        result.append(item)
    return result


@router.get("/notifications/unread-count")
def unread_count(request: Request) -> dict:
    """Returns the total number of unread notifications."""
    store = _get_run_store(request)
    return {"count": store.count_unread()}


@router.post("/notifications/{notification_id}/read")
def mark_read(
    request: Request,
    notification_id: int = Path(..., ge=1),
) -> dict:
    """Marks a single notification as read."""
    store = _get_run_store(request)
    store.mark_read(notification_id)
    return {"status": "ok"}


@router.post("/notifications/read-all")
def mark_all_read(request: Request) -> dict:
    """Marks all unread notifications as read."""
    store = _get_run_store(request)
    store.mark_all_read()
    return {"status": "ok"}
